Highlight tableRows on :hover and :focus-within

BrianLitwin opened the pull request:
Resolves #1041

The purpose of this commit is to make the pagerankTable easier
to read, as it's currently difficult to distinguish which score is
associated with which row because of the tight spacing of the
rows and the space between the score column and the row detail column.

@wchargin provided the implementation using `linearGradient()`
and `backgroundImage`s.

A side effect of highlighting the row on `focus-within` is that the rows
will become highlighted when the expand button is clicked, which we
decided was acceptable.

**Test plan**:

Yarn test passes.

To test the new highlight behavior, visual/manual inspection
passes.

Also added the Aphrodite className to the snapshot
tests. The combination of testing the className + inline style props
should make these tests sensitive to UI changes in the future.

**Screenshots**: 

<img width="939" alt="screenshot 2019-02-17 15 46 34" src="https://user-images.githubusercontent.com/26695477/52918955-332f5280-32cb-11e9-87d3-887c8877116a.png">
<img width="931" alt="screenshot 2019-02-17 15 45 10" src="https://user-images.githubusercontent.com/26695477/52918953-2f9bcb80-32cb-11e9-9356-82c6dccab4ae.png">

wchargin commented on src/explorer/pagerankTable/TableRow.test.js:
```diff
@@ -34,9 +55,29 @@ describe("explorer/pagerankTable/TableRow", () => {
+        expect(expectedClassName._definition.backgroundColor).toBe(
+          backgroundColor
+        );
+      }
```
If none of these cases is hit, the test will silently pass. Someone
could change the loop above from `[0, 1, 2]` to `[0, 1, 2, 3]`, but the
additional invocation wouldn’t test anything with respect to styling.

Two solutions that come to mind:

  - add an `else` clause with `throw new Error("Should be unreachable")`
    or similar;

  - refactor the loop to be over an array of objects like

    ```js
    [
      {depth: 0, expectedAlpha: …},
      …,
    ]
    ```

BrianLitwin commented on src/explorer/pagerankTable/TableRow.js:
```diff
@@ -95,3 +102,17 @@ export function PaddingRow(props: {|+backgroundColor: string|}) {
+});
+
+export function setBackgroundColor(bgColor: string) {
+  return StyleSheet.create({
```
Since aphrodite uses the `className`  prop to set styles, and inline styles uses the`style` prop, I didn't know what the effect would be of mixing them.. how/whether one would take precedence over the other, and how to pass in an effectively empty prop for one when we wanted to use the other method of setting the style. I'll re-think this.

BrianLitwin replied:
Also, worth mentioning for context that the green background is set dynamically, which is why the aphrodite class is created dynamically.

wchargin replied:
Inline styles always take precedence over styles set in stylesheets;
they have maximal specificity. There’s no problem with mixing them.

> Also, worth mentioning for context that the green background is set
> dynamically, which is why the aphrodite class is created dynamically.

Right: you don’t need to do this if you use inline styles, hence the
sggestion.

BrianLitwin commented:
Thanks for the review @wchargin ! 

Do you have an opinion on using React's `onMouseEnter` and `onMouseLeave` functions + inline styles, instead of aphrodite styles? As I'm fiddling w/ it, the downside is having to keep the background color as state in the component.. the upside is throwing out the aphrodite styles, which are tricky when we calculate the background color dynamically. Also, I think testing would be easier/more robust if we could access the `style.backgroundColor` prop, instead of what I'm doing currently to test the aphrodite styling. 

It also seems easier to work with the padding rows in the `TableRow` using inline styles. 

If you suggest that approach, should I create a whole new PR or update this one? It would be almost completely new from a code perspective.  

A UI consideration is what to do with the padding rows. Should they also be highlighted if they are adjacent to the highlighted table row? Otherwise, it looks like: 

<img width="973" alt="screenshot 2019-01-21 16 14 58" src="https://user-images.githubusercontent.com/26695477/51502694-2d665000-1da5-11e9-94fa-218d158133f4.png">

I'm on the fence as to whether that looks acceptable or not. If we highlight the padding rows, however, we might have zero, one, or both highlighted (depending on how many are adjacent), which may look kind of weird/inconsistent.

wchargin commented:
> Do you have an opinion on using React's `onMouseEnter` and
> `onMouseLeave` functions + inline styles, instead of aphrodite styles?

It’s okay to use these where necessary, but I definitely try to avoid
them whenever possible.

For instance, how about using a `linear-gradient` stack?

```jsx
function render() {
  const backgroundColor = `rgba(...)`;
  const makeGradient = (color) => `linear-gradient(to top, ${color}, ${color})`;
  const normalBackground = makeGradient(backgroundColor);
  const highlightBackground = makeGradient("#EBF1F5");
  const backgroundImage = `${normalBackground}, ${highlightBackground}`;
  return (
    <tr style={{backgroundImage}} className={css(styles.hoverHighlight)}>
      {…}
    </tr>
  );
}

const styles = StyleSheet.create({
  hoverHighlight: {
    backgroundSize: "100% 100%, 0 0",
    ":hover": {
      backgroundSize: "100% 100%, 100% 100%",
    },
  },
});
```

Sample fiddle: <https://jsfiddle.net/6y12nws7/>

<details>
<summary>Fiddle source, for redundancy</summary>

HTML:

```html
<table style="border-collapse: collapse">
  <tr class="highlight" style="background-image: linear-gradient(to top, #ff000044, #ff000044), linear-gradient(to top, #0000ff88, #0000ff88)">
    <td>one</td>
    <td>two</td>
  </tr>
  <tr class="highlight" style="background-image: linear-gradient(to top, #ff000044, #ff000044), linear-gradient(to top, #0000ff88, #0000ff88)">
    <td>three</td>
    <td>four</td>
  </tr>
</table>
```

CSS:

```css
tr.highlight {
  background-size: 100% 100%, 0 0;
}
tr.highlight:hover {
  background-size: 100% 100%, 100% 100%;
}
td {
  padding: 6px;
}
```

---

</details>

> should I create a whole new PR or update this one? It would be almost
> completely new from a code perspective.

Prefer keeping the same PR; this keeps the relevant review history in
one place. It’s _totally_ fine to completely change the code as a
consequence of review. I frequently write two or three implementations
of a change before I’m happy with it.

> A UI consideration is what to do with the padding rows. Should they
> also be highlighted if they are adjacent to the highlighted table row?
> Otherwise, it looks like:

Personally, I prefer not highlighting the padding rows, because the
asymmetry would bug me more than the fact that the padding is a
different color.

BrianLitwin commented:
@wchargin Think this is ready for another look. I've mainly ran with the code you provided to set the highlight, which looks good against the green table row background (better than setting the bg directly), but also made this pr much simpler. 

I don't know if I can/ should take the tests any further - currently the tests include snapshot tests for the Aphrodite className and inline style, so I expect the tests would fail in the future if the UI were updated in a way that altered the current behavior.

wchargin commented on src/explorer/pagerankTable/TableRow.js:
```diff
@@ -95,3 +106,12 @@ export function PaddingRow(props: {|+backgroundColor: string|}) {
+const styles = StyleSheet.create({
+  hoverHighlight: {
+    backgroundSize: "100% 100%, 0 0",
+    ":hover, :focus": {
```
This style is applying to the `:focus` pseudo-selector for _every_
element on the page: <https://github.com/Khan/aphrodite/issues/196>.
You can tell because the focus-styling for `<button>` and `<select>`
elements has changed.

You can get the desired _code_ by using two pseudo-selectors…

```diff
diff --git a/src/explorer/pagerankTable/TableRow.js b/src/explorer/pagerankTable/TableRow.js
index 762f80f4..0f38d9df 100644
--- a/src/explorer/pagerankTable/TableRow.js
+++ b/src/explorer/pagerankTable/TableRow.js
@@ -110,7 +110,10 @@ export function PaddingRow(props: {|+backgroundColor: string|}) {
 const styles = StyleSheet.create({
   hoverHighlight: {
     backgroundSize: "100% 100%, 0 0",
-    ":hover, :focus": {
+    ":hover": {
+      backgroundSize: "100% 100%, 100% 100%",
+    },
+    ":focus": {
       backgroundSize: "100% 100%, 100% 100%",
     },
   },
```

…but after testing this I see that it’s not actually achieving the
desired _effect_, because `<tr>`s don’t take focus, so this selector
never applies.

Try it: as you tab through the page, you focus on the various buttons
and links, but the row never gains highlight.

The solution is to use [`:focus-within`][fw] instead of `:focus`.

[fw]: https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within

wchargin commented on src/explorer/pagerankTable/TableRow.js:
```diff
@@ -95,3 +106,12 @@ export function PaddingRow(props: {|+backgroundColor: string|}) {
 }
+
+const styles = StyleSheet.create({
+  hoverHighlight: {
```
Please document the semantics/contract for `hoverHighlight`—it’s not at
all clear from the name (which is fine in this case: there’s not really
a good name that conveys exactly what the client needs to do).

BrianLitwin commented:
The current snag is that 'focus-within' will catch and keep a highlight when you click on the expand button: 

<img width="918" alt="screenshot 2019-02-09 06 43 35" src="https://user-images.githubusercontent.com/26695477/52520301-11343100-2c36-11e9-9f01-ae80c216dc9a.png">
The cursor is over `#286` in the above screenshot. It happens after you click on a button at the third level of indentation. I can't reproduce the issue in IE or Firefox. 

My current solution is [this SO answer] ie adding `onMouseDown={(e)=>e.preventDefault()}` to the button props. 

I can't currently reproduce the thin white bars from the screenshot i originally posted. 

[this SO answer]: https://stackoverflow.com/questions/19053181/how-to-remove-focus-around-buttons-on-click/37580028#37580028

wchargin commented:
> The current snag is that 'focus-within' will catch and keep a
> highlight when you click on the expand button:

Yep. One is focused, and one is hovered. This is the correct behavior,
and you’ll see it across the web (on accessible sites):

  - Navigate to <https://en.wikipedia.org>, tab to a link, and hover
    over another one: both are underlined.

  - Navigate to <https://news.bbc.co.uk>, tab to the “News” nav item,
    and hover the “Sport” nav item: both have their respective colorful
    underlines.

  - Navigate to <https://www.khanacademy.org>, tab to the “Teachers,
    start here” button, and hover the “Parents, start here” button: both
    have a modified text color.

> `onMouseDown={(e)=>e.preventDefault()}`

In addition to removing the focus style, this also prevents focus
entirely. I should be able to click the button and then press the space
bar to click it again.

wchargin commented:
Another one that I just saw:

  - Navigate to <https://www.rust-lang.org/what/cli> and tab to the
    “Rust’s Guarantees” button, then hover the ”Start Configuring”
    button. Both have a white outline.

BrianLitwin commented:
@wchargin So would you approve the behavior as-is or should I be looking for alternatives? Imo the intended behavior was not to highlight a row on button clicks, but it's not a huge trade-off.

wchargin commented:
I think that the current behavior is best out of what’s been mentioned
so far, and I certainly think that this PR is an improvement over the
current state, so I suggest merging as is, after the existing inline is
addressed:
<https://github.com/sourcecred/sourcecred/pull/1059/files#r251710546>

BrianLitwin commented:
@wchargin I've added documentation for the semantics/contract for `hoverOnHighlight`. I updated the commit message too to reflect our changes, with new screenshots.
